Source/Core/Core/IOS/SDIO/SDIOSlot0.cpp

```diff
@@ -187,22 +203,148 @@ s32 SDIOSlot0::ExecuteCommand(const Request& request, u32 _BufferIn, u32 _Buffer
     break;
 
   case SEND_IF_COND:
+    INFO_LOG(IOS_SD, "SEND_IF_COND");
     // If the card can operate on the supplied voltage, the response echoes back the supply
     // voltage and the check pattern that were set in the command argument.
+    // This instruction is used to differentiate between protocol v1 and v2.
+    m_Protocol = PROTOCOL_V2;
     Memory::Write_U32(req.arg, _BufferOut);
     break;
 
   case SEND_CSD:
-    INFO_LOG(IOS_SD, "SEND_CSD");
-    // <WntrMute> shuffle2_, OCR: 0x80ff8000 CID: 0x38a00000 0x480032d5 0x3c608030 0x8803d420
-    // CSD: 0xff928040 0xc93efbcf 0x325f5a83 0x00002600
-
-    // Values used currently are from lpfaint99
-    Memory::Write_U32(0x80168000, _BufferOut);
-    Memory::Write_U32(0xa9ffffff, _BufferOut + 4);
-    Memory::Write_U32(0x325b5a83, _BufferOut + 8);
-    Memory::Write_U32(0x00002e00, _BufferOut + 12);
-    break;
+  {
+    u64 size = m_Card.GetSize();
+    if (m_Protocol == PROTOCOL_V2)
+    {
+      if (size % (512 * 1024) != 0)
+        WARN_LOG(IOS_SD, "SDHC Card size cannot be divided by 1024 * 512");
+
+      size /= 512 * 1024;
+      size -= 1;
+
+      // 0b01               CSD_STRUCTURE (SDv2)
+      // 0b000000           reserved
+      // 0b00001110         TAAC (1.0 * 1ms)
+      // 0b00000000         NSAC
+      // 0b01011010         TRAN_SPEED (5.0 * 10 Mbit/s, max operating frequency)
+
+      // 0b010111110101     CCC (TODO: Figure out what each command class does)
+      // 0b1001             READ_BL_LEN (512 bytes, fixed for SDHC)
+      // 0b0                READ_BL_PARTIAL
+      // 0b0                WRITE_BLK_MISALIGN
+      // 0b0                READ_BLK_MISALIGN
+      // 0b0                DSR_IMP (no driver stage register implemented)
+      // 0b000000           reserved
+      // 0b??????           C_SIZE (most significant 6 bits)
+
+      // 0b???????????????? C_SIZE (least significant 16 bits)
+      // 0b0                reserved
+      // 0b1                ERASE_BLK_EN
+      // 0b1111111          SECTOR_SIZE
+      // 0b0000000          WP_GRP_SIZE (not supported in SDHC)
+
+      // 0b0                WP_GRP_ENABLE
+      // 0b00               reserved
+      // 0b010              R2W_FACTOR (x4)
+      // 0b1001             WRITE_BL_LEN (512 bytes)
+      // 0b0                WRITE_BL_PARTIAL
+      // 0b00000            reserved
+      // 0b0                FILE_FORMAT_GRP
+      // 0b0                COPY
+      // 0b0                PERM_WRITE_PROTECT
+      // 0b0                TMP_WRITE_PROTECT
+      // 0b00               FILE_FORMAT
+      // 0b00               reserved
+      // 0b0000000          CRC
+      // 0b1                reserved
+
+      // TODO: crc7
+      u32 crc = 0;
+
+      Memory::Write_U32(0x400e005a, _BufferOut);
+      Memory::Write_U32(0x5f590000 | (size >> 16), _BufferOut + 4);
+      Memory::Write_U32(0x00007f80 | (size << 16), _BufferOut + 8);
+      Memory::Write_U32(0x0a400001 | crc << 1, _BufferOut + 12);
```

Comment:
I was thinking of a bitfield, but I'm not sure that would make it more readable because honestly, most of these will never have to be dynamic. Especially for v2, most of these fields must have these values (for example, block sizes cannot be anything other than 512 as far as I can tell).
I think it would be cleaner than the separate values though.
Because only three (two for v2) fields are dynamic, I figured the compact version with a descriptive comment would be sufficient.